library/Garden/Web/Route.php

```diff
@@ -0,0 +1,223 @@
+<?php
+/**
+ * @author Todd Burry <todd@vanillaforums.com>
+ * @copyright 2009-2017 Vanilla Forums Inc.
+ * @license Proprietary
+ */
+
+namespace Garden\Web;
+
+
+/**
+ * The base class for routes.
+ *
+ * A route maps {@link RequestInterface} instances to callbacks. A single route will analyze the request and return
+ * dispatch information or **null** if it can't map the route.
+ */
+abstract class Route {
+    const MAP_ARGS = 0x1; // map to path args
+    const MAP_QUERY = 0x2; // map to querystring
+    const MAP_BODY = 0x4; // map to post body
+
+    /**
+     * Route constructor.
+     */
+    public function __construct() {
+        // TODO: Maybe remove this?
+        $this->constraints = [
+            'id' => '`^\d+$`',
+            'slug' => '^[^/.]+$`',
+            'page' => '`^p\d+$`'
+        ];
+    }
+
+    /**
+     * @var array An array of parameter conditions.
+     */
+    private $constraints;
+
+    /**
+     * @var array An array of parameter mappings.
+     */
+    private $mappings = [
+        'query' => Route::MAP_QUERY,
+        'body' => Route::MAP_BODY,
+        'data' => Route::MAP_ARGS | Route::MAP_QUERY | Route::MAP_BODY
+    ];
+
+    /**
+     * Get the conditions.
+     *
+     * @return array Returns the conditions.
+     */
+    public function getConstraints() {
+        return $this->constraints;
+    }
+
+    /**
+     * Set the entire conditions array.
+     *
+     * If you are going to set conditions in this way make sure the array keys are all lowercase.
+     *
+     * @param array $constraints An array of conditions tha maps variable names to **Route::MAP_*** constants.
+     * @return $this
+     */
+    public function setConstraints($constraints) {
```

Comment:
It appears `$constraints` should always be an array. This parameter could probably benefit from a type declaration.